File: mac-app/Tests/QuartzTeachersTests/SiteHealthRepairTests.swift
import XCTest
@testable import QuartzTeachers

/// Putting right the folder problems that can be put right.
@MainActor
final class SiteHealthRepairTests: XCTestCase {

    // MARK: - Functions

    private func makeCourse() throws -> (URL, Course) {
        let root: URL = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("health-repair-\(UUID().uuidString)")
        let courseURL: URL = root.appendingPathComponent("courses/ICS3U")
        try FileManager.default.createDirectory(at: courseURL, withIntermediateDirectories: true)
        let configuration: [String: Any] = [
            "course_code": "ICS3U", "course_name": "Introduction to Computer Science",
            "section_numbers": [1], "num_sections": 1,
            "shared_folders": [], "per_section_folders": ["All Classes"],
            "shared_files": [], "per_section_files": [],
        ]
        let configURL: URL = courseURL.appendingPathComponent("course_config.json")
        try JSONSerialization.data(withJSONObject: configuration, options: [.prettyPrinted])
            .write(to: configURL)
        return (root, Course(
            code: "ICS3U", directoryURL: courseURL,
            configuration: try CourseConfiguration(contentsOf: configURL)
        ))
    }

    private func finding(_ name: String, fixable: Bool, section: Int = 1) -> SiteHealthFinding {
        return SiteHealthFinding(
            name: name, sentence: "Something is wrong.", detail: "More about it.",
            fixable: fixable, course: "ICS3U", section: section
        )
    }

    /// The line that matters most: a fix must restore the FEATURE, not merely
    /// satisfy the check. Recreating an empty curriculum folder would silence
    /// the warning and leave the map missing.
    func testTheUnfixableFindingsAreNeverOffered() {
        for name in ["curriculumCoverageFoundNothing", "courseTeachesNothing",
                     "handWrittenCoveragePage"] {
            XCTAssertFalse(
                SiteHealthRepair.canRepair(finding(name, fixable: true)),
                "\(name) must never grow a fix button, even if the toolchain calls it fixable"
            )
        }
        XCTAssertNil(SiteHealthRepair.buttonTitle(
            for: [finding("curriculumCoverageFoundNothing", fixable: true)]
        ))
    }

    func testAMissingMediaFolderIsPutBack() throws {
        let (root, course) = try makeCourse()
        defer { try? FileManager.default.removeItem(at: root) }

        let media: URL = course.directoryURL.appendingPathComponent("Media")
        XCTAssertFalse(FileManager.default.fileExists(atPath: media.path))

        let repaired = SiteHealthRepair.repair(
            [finding("mediaFolderMissing", fixable: true)], in: course
        )
        XCTAssertEqual(repaired["mediaFolderMissing"], .restored)
        XCTAssertTrue(FileManager.default.fileExists(atPath: media.path))
    }

    func testAMissingFrontPageIsPutBackWithATitle() throws {
        let (root, course) = try makeCourse()
        defer { try? FileManager.default.removeItem(at: root) }

        SiteHealthRepair.repair([finding("sectionIndexMissing", fixable: true)], in: course)

        let index: URL = course.sectionDirectoryURL(forSection: 1)
            .appendingPathComponent("index.md")
        let written: String = try String(contentsOf: index, encoding: .utf8)
        XCTAssertTrue(written.contains("title: Introduction to Computer Science"), written)
    }

    /// A repair whose outcome is invisible is one nobody trusts the second
    /// time — and the Media folder is somewhere a teacher cannot see from this
    /// app, so silence after pressing the button is indistinguishable from
    /// nothing having happened.
    func testARepairSaysWhatItPutBack() {
        XCTAssertEqual(
            SiteHealthRepair.whatWasPutBack(["mediaFolderMissing"]),
            "Put the Media folder back."
        )
        XCTAssertEqual(
            SiteHealthRepair.whatWasPutBack(["mediaFolderMissing", "sectionIndexMissing"]),
            "Put the Media folder and the front page back."
        )
        XCTAssertNil(SiteHealthRepair.whatWasPutBack([]),
                     "nothing repaired means nothing to announce")
        XCTAssertNil(SiteHealthRepair.whatWasPutBack(["curriculumCoverageFoundNothing"]),
                     "a finding with no repair has nothing to report")
    }

    /// The site still shows how things were until it is built again, and saying
    /// so is the difference between a button that fixed the folder and one a
    /// teacher believes fixed their site.
    func testTheTeacherIsToldTheSiteHasNotChangedYet() {
        let said: String = SiteHealthRepair.notOnTheSiteYet
        XCTAssertTrue(said.lowercased().contains("build"), said)
        for word in ["container", "script", "toolchain", "quartz", "config", "rebuild the toolchain"] {
            XCTAssertFalse(said.lowercased().contains(word), "says \"\(word)\" to a teacher")
        }
    }

    /// Pressing Fix twice — or pressing it after putting the folder back in
    /// Obsidian — must not be reported as a permissions problem. Both restores
    /// return "already there", which folding into "failed" turned into
    /// "Plantoir could not put that back… check the folder isn't locked".
    func testNothingToDoIsNotReportedAsAFailure() throws {
        let (root, course) = try makeCourse()
        defer { try? FileManager.default.removeItem(at: root) }

        try FileManager.default.createDirectory(
            at: course.directoryURL.appendingPathComponent("Media"),
            withIntermediateDirectories: true
        )
        let outcome = SiteHealthRepair.outcome(
            ofRepairing: [finding("mediaFolderMissing", fixable: true)], in: course
        )
        XCTAssertEqual(outcome?.headline, "That is already put right.")
        XCTAssertFalse(outcome?.detail.lowercased().contains("read-only") ?? true,
                       "a no-op must not read as a permissions failure")
    }

    /// A PARTIAL failure said nothing about the half that did not come back —
    /// silence whenever anything else succeeded, in the type added so failure
    /// would not be silent.
    func testAPartialFailureNamesWhatDidNotComeBack() throws {
        let (root, course) = try makeCourse()
        defer { try? FileManager.default.removeItem(at: root) }

        // Media can be restored; the front page cannot, because section 9 is
        // not one this course has.
        let outcome = SiteHealthRepair.outcome(
            ofRepairing: [
                finding("mediaFolderMissing", fixable: true),
                finding("sectionIndexMissing", fixable: true, section: 9),
            ],
            in: course
        )
        XCTAssertEqual(outcome?.headline, "Put the Media folder back.")
        XCTAssertTrue(outcome?.detail.contains("Could not put the front page back") ?? false,
                      outcome?.detail ?? "")
        XCTAssertEqual(outcome?.canRebuild, false,
                       "something is still wrong, so do not send them to look at it")
    }

    /// A repair that FAILED must say so. Both restore functions can return
    /// false — a read-only volume, a permissions problem, a file where the
    /// folder should be — and reporting only success made a failed repair
    /// indistinguishable from a successful one: the alert closed either way.
    func testAFailedRepairIsReportedRatherThanPassedOverInSilence() throws {
        let (root, course) = try makeCourse()
        defer { try? FileManager.default.removeItem(at: root) }

        // A FILE where the Media folder should be: creating the directory
        // cannot succeed.
        try "not a folder".write(
            to: course.directoryURL.appendingPathComponent("Media"),
            atomically: true, encoding: .utf8
        )

        let outcome = SiteHealthRepair.outcome(
            ofRepairing: [finding("mediaFolderMissing", fixable: true)], in: course
        )
        XCTAssertNotNil(outcome, "a failed repair must still produce something to show")
        XCTAssertTrue(outcome?.headline.contains("could not") ?? false, outcome?.headline ?? "")
        XCTAssertEqual(outcome?.canRebuild, false,
                       "there is nothing to see, so do not offer to build")
    }

    /// A fresh preview is no use to somebody whose site is published: only
    /// publishing again changes what students look at.
    func testAPublishedSiteIsNotOfferedAPreview() throws {
        let (root, course) = try makeCourse()
        defer { try? FileManager.default.removeItem(at: root) }

        let outcome = SiteHealthRepair.outcome(
            ofRepairing: [finding("mediaFolderMissing", fixable: true)], in: course,
            occasion: .publishing
        )
        XCTAssertEqual(outcome?.canRebuild, false)
        XCTAssertTrue(outcome?.detail.lowercased().contains("publish") ?? false,
                      outcome?.detail ?? "")
    }

    /// A finding's section number is parsed from output and falls back to 0.
    /// Creating a `section0` folder because a line was malformed would invent
    /// structure the course does not have.
    func testARepairRefusesASectionTheCourseDoesNotHave() throws {
        let (root, course) = try makeCourse()
        defer { try? FileManager.default.removeItem(at: root) }

        XCTAssertFalse(SiteHealthRepair.restoreSectionIndex(forSection: 0, in: course))
        XCTAssertFalse(SiteHealthRepair.restoreSectionIndex(forSection: 7, in: course))
        XCTAssertFalse(
            FileManager.default.fileExists(
                atPath: course.directoryURL.appendingPathComponent("section0").path
            ),
            "no section0 folder may be conjured up"
        )
    }

    /// Pressing it twice, or pressing it after fixing the problem in Obsidian,
    /// must change nothing — a repair that overwrote would destroy the very
    /// page the teacher had just written.
    func testARepairNeverOverwritesWhatIsAlreadyThere() throws {
        let (root, course) = try makeCourse()
        defer { try? FileManager.default.removeItem(at: root) }

        let index: URL = course.sectionDirectoryURL(forSection: 1)
            .appendingPathComponent("index.md")
        try FileManager.default.createDirectory(
            at: index.deletingLastPathComponent(), withIntermediateDirectories: true
        )
        try "---\ntitle: My own front page\n---\nWelcome!\n"
            .write(to: index, atomically: true, encoding: .utf8)

        let repaired = SiteHealthRepair.repair(
            [finding("sectionIndexMissing", fixable: true)], in: course
        )
        XCTAssertEqual(repaired["sectionIndexMissing"], .alreadyFine,
                       "already there is not a failure")
        XCTAssertTrue(
            try String(contentsOf: index, encoding: .utf8).contains("Welcome!"),
            "the teacher's own page must survive"
        )
    }
}

File: windows-app/Plantoir.Tests/SiteHealthRepairTests.cs
using System.Text;
using Plantoir.Core.Models;
using Plantoir.Core.Scripting;

namespace Plantoir.Tests;

/// <summary>
/// Putting a folder problem right: what is offered, what is refused, and what
/// the teacher is told afterwards.
///
/// <para>The tests that matter here are the ones about NOT doing something — a
/// teacher's own front page surviving a repair, and a check with no real repair
/// never acquiring a button. A repair that satisfies its check without
/// restoring the feature is worse than no repair, because the warning goes away
/// and the site stays wrong.</para>
/// </summary>
/// <remarks>
/// In the serialized collection because a successful repair writes to the
/// activity trail, whose log path is a process-wide static.
/// </remarks>
[Collection(SharedActivityState.Name)]
public class SiteHealthRepairTests : IDisposable
{
    private readonly string _trailPath;
    private readonly string _root;

    public SiteHealthRepairTests()
    {
        _trailPath = Path.Combine(Path.GetTempPath(), $"plantoir-trail-{Guid.NewGuid():N}.txt");
        ActivityTrail.SetCustomLogPathForTesting(_trailPath);
        _root = Path.Combine(Path.GetTempPath(), $"plantoir-repair-{Guid.NewGuid():N}");
    }

    public void Dispose()
    {
        ActivityTrail.SetCustomLogPathForTesting(TestTrailRedirect.ScratchTrailPath);
        try { File.Delete(_trailPath); } catch { }
        try { Directory.Delete(_root, recursive: true); } catch { }
    }

    private string Trail() => File.Exists(_trailPath) ? File.ReadAllText(_trailPath) : "";

    private const string ConfigJson =
        "{\"course_code\": \"ICS3U\", \"course_name\": \"Introduction to Computer Science\", " +
        "\"section_numbers\": [1, 2], \"num_sections\": 2}";

    private Course MakeCourse()
    {
        string courseDir = Path.Combine(_root, "courses", "ICS3U");
        Directory.CreateDirectory(courseDir);
        File.WriteAllText(Path.Combine(courseDir, "course_config.json"), ConfigJson);
        return new Course("ICS3U", courseDir,
                          CourseConfiguration.FromBytes(Encoding.UTF8.GetBytes(ConfigJson)));
    }

    private static SiteHealthFinding Finding(string name, bool fixable = true, int section = 1) =>
        new(name, "A sentence.", "Some detail.", fixable, "ICS3U", section);

    // ---- What may be offered --------------------------------------------

    [Theory]
    [InlineData("curriculumCoverageFoundNothing")]
    [InlineData("courseTeachesNothing")]
    [InlineData("handWrittenCoveragePage")]
    [InlineData("noGradedFolders")]
    public void ACheckWithNoRealRepairIsNeverOffered(string name)
    {
        // The point of the whole design: recreating an empty curriculum folder
        // would silence the warning and leave the map missing.
        var finding = Finding(name, fixable: true);
        Assert.False(SiteHealthRepair.CanRepair(finding));
        Assert.Null(SiteHealthRepair.ButtonTitle(new[] { finding }));
    }

    [Fact]
    public void TheFlagAloneIsNotEnoughAndNeitherIsTheNameAlone()
    {
        // The flag means "this kind of thing is repairable"; the name is what
        // says THIS app has a repair. Both, or no button.
        Assert.True(SiteHealthRepair.CanRepair(Finding("mediaFolderMissing", fixable: true)));
        Assert.False(SiteHealthRepair.CanRepair(Finding("mediaFolderMissing", fixable: false)));
        Assert.False(SiteHealthRepair.CanRepair(Finding("noGradedFolders", fixable: true)));
    }

    [Fact]
    public void TheButtonNamesTheOneThingItWillDo()
    {
        Assert.Equal("Put the Media folder back",
            SiteHealthRepair.ButtonTitle(new[] { Finding("mediaFolderMissing") }));
        Assert.Equal("Add the missing page",
            SiteHealthRepair.ButtonTitle(new[] { Finding("sectionIndexMissing") }));
        Assert.Equal("Put them back",
            SiteHealthRepair.ButtonTitle(new[] { Finding("mediaFolderMissing"), Finding("sectionIndexMissing") }));
    }

    // ---- Never overwrite -------------------------------------------------

    [Fact]
    public void ATeachersOwnFrontPageSurvivesARepair()
    {
        // The test that matters. A repair that overwrote this would destroy the
        // teacher's work in the act of "helping".
        var course = MakeCourse();
        string sectionDir = course.SectionDirectory(1);
        Directory.CreateDirectory(sectionDir);
        string index = Path.Combine(sectionDir, "index.md");
        const string theirs = "---\ntitle: Welcome to my class\n---\n\nRead this first.\n";
        File.WriteAllText(index, theirs);

        var result = SiteHealthRepair.RestoreIndex(1, course);

        Assert.Equal(SiteHealthRepair.Result.AlreadyFine, result);
        Assert.Equal(theirs, File.ReadAllText(index));
        // Nobody acted, so nothing is claimed on the trail.
        Assert.DoesNotContain("put the front page back", Trail());
    }

    [Fact]
    public void AMediaFolderThatIsAlreadyThereIsLeftAloneAndIsNotAFailure()
    {
        var course = MakeCourse();
        string media = Path.Combine(course.DirectoryPath, "Media");
        Directory.CreateDirectory(media);
        File.WriteAllText(Path.Combine(media, "diagram.png"), "not really a png");

        Assert.Equal(SiteHealthRepair.Result.AlreadyFine, SiteHealthRepair.RestoreMedia(course));
        Assert.True(File.Exists(Path.Combine(media, "diagram.png")));
    }

    // ---- Doing it --------------------------------------------------------

    [Fact]
    public void TheRestoredFrontPageCarriesTheCourseNameAndNothingElse()
    {
        var course = MakeCourse();

        Assert.Equal(SiteHealthRepair.Result.Restored, SiteHealthRepair.RestoreIndex(2, course));

        string index = Path.Combine(course.SectionDirectory(2), "index.md");
        // Character for character with the mac: the same button on the two
        // platforms must not write differently titled pages into a vault.
        Assert.Equal("---\ntitle: Introduction to Computer Science\n---\n", File.ReadAllText(index));
        // And it lands in the section it was asked about, not just somewhere.
        Assert.False(File.Exists(Path.Combine(course.SectionDirectory(1), "index.md")));
        Assert.Contains("put the front page back", Trail());
        Assert.Contains("ICS3U/2", Trail());
    }

    [Fact]
    public void TheRestoredMediaFolderIsEmptyAndIsRecordedAgainstTheCourse()
    {
        var course = MakeCourse();

        Assert.Equal(SiteHealthRepair.Result.Restored, SiteHealthRepair.RestoreMedia(course));

        string media = Path.Combine(course.DirectoryPath, "Media");
        Assert.True(Directory.Exists(media));
        Assert.Empty(Directory.GetFileSystemEntries(media));
        Assert.Contains("put the Media folder back", Trail());
        // Media belongs to the whole course; a section number here would name
        // a section the finding may not even be about.
        Assert.DoesNotContain("ICS3U/1 ·", Trail());
    }

    // ---- Refusing rather than inventing structure ------------------------

    [Fact]
    public void AMalformedLineCannotConjureASectionZeroFolder()
    {
        // SiteHealthFinding.Parse falls back to section 0 when the field is
        // missing or the wrong type. Creating section0 would invent structure
        // the course does not have.
        var course = MakeCourse();

        Assert.Equal(SiteHealthRepair.Result.Failed, SiteHealthRepair.RestoreIndex(0, course));
        Assert.False(Directory.Exists(Path.Combine(course.DirectoryPath, "section0")));
    }

    [Fact]
    public void ADirectorySittingWhereTheFrontPageBelongsIsAFailureNotAnAlreadyFine()
    {
        // File.Exists answers false for a directory, so without an explicit
        // check the write below fails and the teacher is told to see whether
        // their disk is read-only — which is not the problem. The mac reports
        // this one as ALREADY PUT RIGHT, which is worse: the site still has no
        // front page. See MAC-HANDOFF.
        var course = MakeCourse();
        Directory.CreateDirectory(Path.Combine(course.SectionDirectory(1), "index.md"));

        Assert.Equal(SiteHealthRepair.Result.Failed, SiteHealthRepair.RestoreIndex(1, course));
    }

    [Fact]
    public void TwoRepairsAreNamedInOneSentence()
    {
        var course = MakeCourse();
        var outcome = SiteHealthRepair.OutcomeOfRepairing(
            new[] { Finding("mediaFolderMissing"), Finding("sectionIndexMissing") }, course);

        Assert.Equal("Put the Media folder and the front page back.", outcome!.Headline);
        Assert.Equal(SiteHealthRepair.NotOnTheSiteYet, outcome.Detail);
        Assert.True(outcome.CanRebuild);
    }

    [Fact]
    public void TwoSectionsMissingAFrontPageDoNotCollapseIntoOneAnswer()
    {
        // Both are repaired either way; what a name-keyed result loses is the
        // REPORT. Section 1 restored and section 2 already fine would come back
        // as "that is already put right", with no preview offered, because the
        // second result overwrote the first.
        var course = MakeCourse();
        Directory.CreateDirectory(course.SectionDirectory(2));
        File.WriteAllText(Path.Combine(course.SectionDirectory(2), "index.md"), "theirs");

        var outcome = SiteHealthRepair.OutcomeOfRepairing(
            new[] { Finding("sectionIndexMissing", section: 1), Finding("sectionIndexMissing", section: 2) },
            course);

        Assert.Equal("Put the front page back.", outcome!.Headline);
        Assert.True(outcome.CanRebuild);
        Assert.True(File.Exists(Path.Combine(course.SectionDirectory(1), "index.md")));
        Assert.Equal("theirs", File.ReadAllText(Path.Combine(course.SectionDirectory(2), "index.md")));
    }

    [Fact]
    public void WhenEverythingFailedTheHeadlineDoesNotTryToNameThem()
    {
        // Deliberate: with nothing put back there is no "Put X back." sentence
        // to lead with, and the explanation is the same for either. The helper
        // that DOES name two is pinned beside it, because the two-name join is
        // otherwise only exercised on the success path.
        var course = MakeCourse();
        File.WriteAllText(Path.Combine(course.DirectoryPath, "Media"), "a file, not a folder");

        var outcome = SiteHealthRepair.OutcomeOfRepairing(
            new[] { Finding("mediaFolderMissing"), Finding("sectionIndexMissing", section: 9) }, course);

        Assert.Equal("Plantoir could not put that back.", outcome!.Headline);
        Assert.Equal(SiteHealthRepair.CouldNotExplanation, outcome.Detail);
        Assert.False(outcome.CanRebuild);

        Assert.Equal("Could not put the Media folder and the front page back.",
            SiteHealthRepair.WhatCouldNotBePutBack(new[] { "mediaFolderMissing", "sectionIndexMissing" }));
    }

    [Fact]
    public void AFileSittingWhereTheMediaFolderBelongsIsAFailureNotAnAlreadyFine()
    {
        var course = MakeCourse();
        File.WriteAllText(Path.Combine(course.DirectoryPath, "Media"), "a file, not a folder");

        Assert.Equal(SiteHealthRepair.Result.Failed, SiteHealthRepair.RestoreMedia(course));
    }

    // ---- What the teacher is told ---------------------------------------

    [Fact]
    public void NothingRepairableMeansNothingIsReported()
    {
        var course = MakeCourse();
        Assert.Null(SiteHealthRepair.OutcomeOfRepairing(
            new[] { Finding("curriculumCoverageFoundNothing") }, course));
    }

    [Fact]
    public void PressingTwiceReadsAsAlreadyPutRightRatherThanAsAPermissionsProblem()
    {
        var course = MakeCourse();
        var findings = new[] { Finding("mediaFolderMissing") };

        var first = SiteHealthRepair.OutcomeOfRepairing(findings, course);
        var second = SiteHealthRepair.OutcomeOfRepairing(findings, course);

        Assert.Equal("Put the Media folder back.", first!.Headline);
        Assert.Equal("That is already put right.", second!.Headline);
        Assert.Equal("Nothing needed changing.", second.Detail);
        // Nothing changed, so there is nothing to go and look at.
        Assert.False(second.CanRebuild);
    }

    [Fact]
    public void APartialFailureNamesTheHalfThatDidNotComeBack()
    {
        // Media can be restored; the front page cannot, because section 9 is
        // not a section of this course.
        var course = MakeCourse();
        var outcome = SiteHealthRepair.OutcomeOfRepairing(
            new[] { Finding("mediaFolderMissing"), Finding("sectionIndexMissing", section: 9) }, course);

        Assert.Equal("Put the Media folder back.", outcome!.Headline);
        Assert.Contains("Could not put the front page back.", outcome.Detail);
        Assert.Contains(SiteHealthRepair.CouldNotExplanation, outcome.Detail);
        // Something did not come back, so "Preview Again" would be showing the
        // teacher a site that is still wrong.
        Assert.False(outcome.CanRebuild);
    }

    [Fact]
    public void AFailureOnItsOwnIsReportedRatherThanClosingSilently()
    {
        var course = MakeCourse();
        var outcome = SiteHealthRepair.OutcomeOfRepairing(
            new[] { Finding("sectionIndexMissing", section: 9) }, course);

        Assert.Equal("Plantoir could not put that back.", outcome!.Headline);
        Assert.Equal(SiteHealthRepair.CouldNotExplanation, outcome.Detail);
        Assert.False(outcome.CanRebuild);
    }

    [Fact]
    public void TheOccasionChangesTheSentenceAndNotTheOffer()
    {
        var building = SiteHealthRepair.OutcomeOfRepairing(
            new[] { Finding("mediaFolderMissing") }, MakeCourse(), SiteHealthRepair.Occasion.Building);

        // A second course, so the first one's repair does not make this one a
        // no-op.
        string otherRoot = Path.Combine(_root, "other");
        Directory.CreateDirectory(Path.Combine(otherRoot, "courses", "ICS3U"));
        var otherCourse = new Course("ICS3U", Path.Combine(otherRoot, "courses", "ICS3U"),
                                     CourseConfiguration.FromBytes(Encoding.UTF8.GetBytes(ConfigJson)));
        var publishing = SiteHealthRepair.OutcomeOfRepairing(
            new[] { Finding("mediaFolderMissing") }, otherCourse, SiteHealthRepair.Occasion.Publishing);

        Assert.Equal(SiteHealthRepair.NotOnTheSiteYet, building!.Detail);
        Assert.Equal(SiteHealthRepair.NotPublishedYet, publishing!.Detail);
        // The preview is offered on BOTH occasions — the reversal recorded in
        // shared-rules.json -> siteHealth.repair.afterwards.
        Assert.True(building.CanRebuild);
        Assert.True(publishing.CanRebuild);
    }

    [Fact]
    public void ThePublishSentenceDoesNotAssertAPublishThatMayNeverHaveHappened()
    {
        // It is shown after a FAILED deploy too, and for a section publishing
        // for the first time nothing has ever gone out.
        Assert.DoesNotContain("last published", SiteHealthRepair.NotPublishedYet);
        Assert.DoesNotContain("students still see", SiteHealthRepair.NotPublishedYet);
        Assert.Contains("until you publish again", SiteHealthRepair.NotPublishedYet);
    }
}
